### ISSUE TITLE
Offer untopped queues

### ISSUE THREAD

**Issue by treeowl** [lspitzner/pqueue#33]
`BinomHeap` can easily be wrapped up in a `newtype` to offer basic binomial queues that don't track size and don't store the minimal entry separately. For some applications, such as heapsort, this should be significantly more efficient. Let's do it.

**Comment by treeowl** [comment]
I've done this now (for plain min-queues); there were no real difficulties. But as it is, there's a lot of code duplication. The fix, I think, is to define

```haskell
data MinQueue a = Empty | MinQueue {-# UNPACK #-} !Int !a !(BMinQueue a)
```

where `BMinQueue` is the "untopped" queue. For certain purposes, we'll need a conversion function from `BMinQueue` to `MinQueue` that extracts the minimal element and calculates the size. We could do those together, but I'm not sure it's worth the trouble. Even without that, I think a number of functions (partitions, say) will get faster, because it's more efficient to build a large untopped queue and then extract its minimum than to build a large topped queue. O(n) fewer comparisons, and probably also O(n) less allocation.

**Comment by konsumlamm** [comment]
Hmm, I'm really not sure what to think of this one. On one hand, more efficient functions sound good, but otoh I'd really like to avoid the code duplication.

I don't think the average `pqueue` user would use such an untopped queue, nor do I think that heapsort is a common application (there are better sorting algorithms). So maybe this should be in an internal module or a separate package? However, that still doesn't avoid the code duplication...

**Comment by konsumlamm** [comment]
Isn't a `BMinQueue` just a `BinomHeap` anyway? So we don't really need new types/modules if we just want more efficient operations.

**Comment by konsumlamm** [comment]
Isn't a `BMinQueue` just a `BinomHeap` anyway? So we don't really need new types/modules if we just want more efficient operations.

**Comment by treeowl** [comment]
The code duplication is gone. That was just an intermediate stage. I think the overall situation is much better now. Wouldn't this untopped version also be better for shortest path algorithms?

**Comment by treeowl** [comment]
> Isn't a `BMinQueue` just a `BinomHeap` anyway? So we don't really need new types/modules if we just want more efficient operations.

Yes, a `BMinQueue` is a `BinomHeap`. It's an actual binomial queue, rather than one augmented with extra stuff. But giving it a newtype wrapper has at least two big benefits:

1. It can be treated as a proper abstract datatype.
2. We can give it typeclass instances.

I think these are both necessary and sufficient for principled use outside the package. As you can see from the PR, (1) is also quite useful *within* the package, where it lets us write `compare` much more succinctly.

To repeat what I said above, the untopped queues aren't just for sorting. They're for pretty much any algorithm that doesn't need to be able to peek at the minimal element in O(1) time and does not see mostly decreasing keys. This includes, at least, shortest path algorithms and k-way merge algorithms. Untopped queues are particularly beneficial when comparisons are not extremely cheap.

**Comment by konsumlamm** [comment]
> The code duplication is gone. That was just an intermediate stage. I think the overall situation is much better now.

Fair, it is better now. However, we still have twice the functions (that we need to maintain). While I don't see a way to avoid this, I still don't like it.

> Wouldn't this untopped version also be better for shortest path algorithms?

Those usually also use "decrease priority" operations (at least Dijkstra's algorithm does) which our priority queues don't provide, so I'm not sure how those would best be implemented using `pqueue` anyway. But apart from that you're right.

> But giving it a newtype wrapper has at least two big benefits:

Makes sense.

> To repeat what I said above, the untopped queues aren't just for sorting.

I just mentioned heapsort because you listed it as an argument for this change.

> They're for pretty much any algorithm that doesn't need to be able to peek at the minimal element in O(1) time and does not see mostly decreasing keys. This includes, at least, shortest path algorithms and k-way merge algorithms. Untopped queues are particularly beneficial when comparisons are not extremely cheap.

Yeah, I think many algorithms would benefit from this (at least in theory), but we should have benchmarks to verify that (#48). One problem I see is that with untopped/bare queues, it may be hard for the users to decide which version they want (perhaps you don't know beforehand whether you'll need O(1) `size`/`getMin`, although ig you can just switch out the module).

**Comment by treeowl** [comment]
Since people might well want to experiment with whether they want bare queues or augmented queues, it might be better to give the types the same names. I'm thinking I should rename `Data.PQueue.Bare` to `Data.BinomialQueue` (or just `BinomialQueue`? That seems to be the current fashion) and make the types `MinQueue` and so on.

**Comment by treeowl** [comment]
For reasons I really, really don't understand, the bare queues don't do well on the k-way merge benchmark. What am I missing?

**Comment by treeowl** [comment]
@konsumlamm , assuming the corrected benchmarks like this branch, should I go with `Data.BinomQueue` or the more modern `BinomQueue`?

**Comment by konsumlamm** [comment]
Hmm, I'm not really sure. Where have you got the impression that the modern naming convention is to drop the `Control`/`Data` prefix (just curious, I haven't heard of this trend)? I would actually prefer not having a `Data` prefix (maybe a `PQueue` prefix, since that's the package name?), but isn't it a bit late for that? As for `BinomQueue` vs `BinomialQueue`, I prefer the latter, it's not really worth it to abbreviate "binomial" here imo.

Do you have any solution to the problem of doubling the surface API? I could imagine a tool that checks that the bare variants expose the same interface. (Most) documentation changes would then need to change 8 doc comments instead of 4, although the comments aren't necessarily equal.

**Comment by treeowl** [comment]
I don't know of such a tool, but I don't think that should be considered a blocker. `witherable` moved from `Data.Witherable` to `Witherable`. It's not really too late; we just need a deprecation cycle so people can change their imports. I'm okay with `BinomialQueue`.

**Comment by treeowl** [comment]
Whatever we do, I want to do it fast... the benchmarks reveal that the new deletion code in #26 led to perf regression. I'd like to chase that down, and also look at getting specialization, but keeping the `bare` branch up to date through any of that will be very painful.

**Comment by konsumlamm** [comment]
Let's go with `BinomialQueue`.

**Comment by treeowl** [comment]
One option to check for the interfaces being similar is to have a test that dumps out all the exported names from each module (I'm pretty sure Template Haskell can do this). Then it can output a diff, and fail if the diff is not what we expect it to be.
